### PR TITLE
Hide disabled API roles when selecting roles for participant

### PR DESCRIPTION
**Pull request by joshua-smee-ttd**
Added migration which adds disabled field to apiRoles in DB. Disabled all keys except for known good keys by default.

image
DB table
image
Resulting approval screen

### CODE REVIEW

File: src/database/migrations/20240104050829_DisableOptOutRole.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { Knex } from 'knex';
+
+export async function up(knex: Knex): Promise<void> {
+  await knex.schema.alterTable('apiRoles', (table) => {
+    table.boolean('disabled').defaultTo(true).notNullable();
+  });
+
+  const allowedRoles = ['MAPPER', 'GENERATOR', 'ID_READER', 'SHARER'];
```
Comment (alex-yau-ttd) [inline src/database/migrations/20240104050829_DisableOptOutRole.ts:8]: perhaps `enabledRoles` makes more sense here?

File: src/api/entities/ApiRole.ts [hunk 2]
```diff
@@ -25,12 +25,14 @@ export class ApiRole extends BaseModel {
...
 }
 
-export type ApiRoleDTO = ModelObjectOpt<ApiRole>;
+export type ApiRoleDTO = Omit<ModelObjectOpt<ApiRole>, 'disabled'>;
 
 export const ApiRoleSchema = z.object({
   id: z.number(),
   roleName: z.string(),
   externalName: z.string(),
+  disabled: z.boolean(),
```
Comment (alex-yau-ttd) [inline src/api/entities/ApiRole.ts:37]: I don't think we need this line?

File: src/api/configureApi.ts [hunk 3]
```diff
@@ -159,7 +159,7 @@ export function configureAndStartApi(useMetrics: boolean = true) {
   });
 
   router.get('/apiRoles', async (_req, res) => {
-    const apiRoles = await ApiRole.query();
+    const apiRoles = await ApiRole.query().where({ disabled: false });
```
Comment (alex-yau-ttd) [inline src/api/configureApi.ts:162]: can this just be `.where('disabled', false })`?

File: src/api/entities/ApiRole.ts (hunk 2 above)
Comment (joshua-smee-ttd) [reply src/api/entities/ApiRole.ts:37]: Yeah we don't atm. I would think though it's better to have for later? Happy to remove if not really a point.

File: src/database/migrations/20240104050829_DisableOptOutRole.ts (hunk 1 above)
Comment (joshua-smee-ttd) [reply src/database/migrations/20240104050829_DisableOptOutRole.ts:8]: Ah yeah that does work better.

File: src/api/entities/ApiRole.ts (hunk 2 above)
Comment (alex-yau-ttd) [reply src/api/entities/ApiRole.ts:37]: I think remove it if we aren't using it now. We may never use it?

Comment (joshua-smee-ttd) [reply src/api/entities/ApiRole.ts:37]: Sounds good!

**Review by alex-yau-ttd — APPROVED** [review]

**Pull request merged**